vm/devices/storage/scsidisk/src/lib.rs

```diff
@@ -867,6 +885,134 @@ impl SimpleScsiDisk {
+        // Validate atomic-specific parameters
+        // If atomic boundary is 0, make sure transfer length not larger than max atomic
+        // transfer length
+        if atomic_boundary == 0
+            && len as u32
+                > self
+                    .scsi_parameters
+                    .maximum_atomic_transfer_length
+                    .unwrap_or(0)
+        {
+            tracelimit::error_ratelimited!(
+                len,
+                maximum_atomic_transfer_length = self
+                    .scsi_parameters
+                    .maximum_atomic_transfer_length
+                    .unwrap_or(0),
+                "atomic boundary 0, transfer length > maximum atomic transfer length"
+            );
+            return Err(ScsiError::IllegalRequest(AdditionalSenseCode::INVALID_CDB));
+        }
+
+        // If atomic boundary greater than 0, make sure transfer length not larger than max atomic
+        // transfer length with atomic boundary
+        if atomic_boundary > 0
+            && len as u32
+                > self
+                    .scsi_parameters
+                    .maximum_atomic_transfer_length_with_atomic_boundary
+                    .unwrap_or(0)
+        {
+            tracelimit::error_ratelimited!(
+                atomic_boundary,
+                maximum_atomic_transfer_length_with_atomic_boundary = self
+                    .scsi_parameters
+                    .maximum_atomic_transfer_length_with_atomic_boundary
+                    .unwrap_or(0),
+                "atomic boundary > 0, transfer length > maximum atomic transfer length with atomic boundary"
+            );
+            return Err(ScsiError::IllegalRequest(AdditionalSenseCode::INVALID_CDB));
+        }
+
+        // If atomic boundary non-zero, make sure less than or equal to max atomic boundary size
+        if atomic_boundary > 0
+            && atomic_boundary as u32
+                > self
+                    .scsi_parameters
+                    .maximum_atomic_boundary_size
+                    .unwrap_or(0)
+        {
+            tracelimit::error_ratelimited!(
+                atomic_boundary,
+                maximum_atomic_boundary_size = self
+                    .scsi_parameters
+                    .maximum_atomic_boundary_size
+                    .unwrap_or(0),
+                "atomic boundary > maximum atomic boundary size"
+            );
+            return Err(ScsiError::IllegalRequest(AdditionalSenseCode::INVALID_CDB));
+        }
+
+        // If atomic transfer length granularity non-zero, make sure that either transfer length OR
+        // the non-zero atomic boundary field is a multiple of this granularity
+        if self
+            .scsi_parameters
+            .atomic_transfer_length_granularity
+            .unwrap_or(0)
+            > 0
+        {
+            if len as u32
+                % self
+                    .scsi_parameters
+                    .atomic_transfer_length_granularity
```

Comment:
Is it possible to require the granularity to be a power of 2? If so, avoid `%`--store a shift and use shift/mask operations.